rust/libnewsboat/src/filterparser.rs

```diff
@@ -0,0 +1,287 @@
+extern crate nom;
+
+use std::io::{self, Read, BufRead};
+use nom::*;
+use nom::types::CompleteStr;
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum Operation {
+    Equals,
+    NotEquals,
+    RegEquals,
+    NotRegEquals,
+    LessThan,
+    GreaterThan,
+    LessThanOrEquals,
+    GreaterThanOrEquals,
+    Between,
+    Contains,
+    NotContains,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum Value {
+    Str(String),
+    Int(i32),
+    Range(i32, i32)
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Comparison {
+    pub attribute: String,
+    pub op: Operation,
+    pub value: Value
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum Expression {
+    And(Box<Expression>, Box<Expression>),
+    Or(Box<Expression>, Box<Expression>),
+    Comparison(Comparison)
+}
+
+named!(operators<CompleteStr, Operation>,
+    do_parse!(
+        tag: alt!(
+            map!(tag!("=="), |_| Operation::Equals) |
+            map!(tag!("!="), |_| Operation::NotEquals) |
+            map!(tag!("=~"), |_| Operation::RegEquals) |
+            map!(tag!("=") , |_| Operation::Equals) |
+            map!(tag!("!~"), |_| Operation::NotRegEquals) |
+            map!(tag!("<="), |_| Operation::LessThanOrEquals) |
+            map!(tag!(">="), |_| Operation::GreaterThanOrEquals) |
+            map!(tag!("<") , |_| Operation::LessThan) |
+            map!(tag!(">") , |_| Operation::GreaterThan) |
+            map!(tag!("between"), |_| Operation::Between) |
+            map!(tag!("#"), |_| Operation::Contains) |
+            map!(tag!("!#"), |_| Operation::NotContains)
+        ) >>
+        (tag)
+    )
+);
+
+fn quoted_string(input: CompleteStr) -> IResult<CompleteStr, Value>{
+    do_parse!(input,
+        tag!("\"")
+        >> value: escaped_transform!(
+                    is_not!("\\\""),
+                    '\\',
+                    alt!(
+                        tag!("\\") => { |_| &"\\"[..] } |
+                        tag!("\"") => { |_| &"\""[..] }
+                    )
+            )
+        >> tag!("\"")
+        >> (value)
+    ).map(|result| {
+        (result.0, Value::Str(result.1))
+    })
+}
+
+fn number(input: CompleteStr) -> IResult<CompleteStr, Value> {
+    do_parse!(input,
+        num: recognize!(
+            do_parse!(
+                opt!(tag!("-")) >>
+                take_while1!(|c| is_digit(c as u8)) >>
+                ()
+            )
+        ) >> (num)
+    ).map(|result| {
+        let tuple = result.1;
+
+        let number: i32 = tuple.0.parse().unwrap();
```

Comment:
Does unwrap provides an error handling here?

Reply:
It doesn't, if it's invalid newsboat crashes with this code. However, the filter right above this should make this impossible as it pretty much guarantees it's a valid number.

The only scenario I can see this failing is if a number is so large is doesn't fit within the i32, I'm not sure what the behaviour of `parse()` is in that case.

Reply:
I think `parse` will break if the number is too large too.

Reply:
Yeah, [it'll return an error](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=83e89e7bc006be95ad17e777f2f87993).